setup - add lvs subkey to ansible_facts['vgs'] values

s-hertel opened the pull request:
##### SUMMARY

In triage we discussed deprecating ansible_facts['lvs'], which can be incomplete, in favor of extracting the logical volumes from ansible_facts['vgs'], but upon closer inspection, this data isn't available there yet. This adds that.

Fixes #85632

##### ISSUE TYPE

- Feature Pull Request

s-hertel committed:
Add lvs subkey to ansible_facts['vgs'] dictionary values

Add a note to consider deprecating ansible_facts['lvs'] in favor of ansible_facts['vgs']

bcoca commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -901,7 +902,13 @@ def get_lvm_facts(self):
+                    lv_name = items[0]
+                    # The LV name is only unique per VG, so the top level fact lvs can be misleading.
+                    # TODO: deprecate lvs in favor of vgs
+                    lvs[lv_name] = {'size_g': items[3], 'vg': vg_name}
```
Not sure if this is working all the way back to controller, but the code would be something like this:
```
from ansible.module_utils._internal import  _deprecator
from ansible.module_utils._internal._datatag import _tags

_DEPRECATE_LVS = _tags.Deprecated(
    msg='The internal "lvs" fact is deprecated.',
    version='2.24',
    deprecator=_deprecator.ANSIBLE_CORE_DEPRECATOR,
    help_text='Use the `lvs`  entry under each volume group instead.',
)
```

then
```
 lvs[lv_name] = _DEPRECATE_LVS.tag({'size_g': items[3], 'vg': vg_name})
```

s-hertel replied:
I was thinking we'd need to add the alternative at least a release before deprecating. Is that not true?

bcoca replied:
did we ever make that a rule? still, it is a much nicer way to give people a transition, so agreed.

bcoca commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -901,7 +902,13 @@ def get_lvm_facts(self):
+                    lvs[lv_name] = {'size_g': items[3], 'vg': vg_name}
+                    if vg_name in vgs:
+                        vgs[vg_name]['lvs'][lv_name] = {'size_g': items[3]}
 
```
we should get an error warning if not in vgs? this means either it changed while we were gathering or that we missed adding it.

s-hertel replied:
I added a warning (and changed this to a try/except KeyError).

bcoca commented on test/integration/targets/hardware_facts/tasks/Linux.yml:
```diff
@@ -52,7 +52,22 @@
+          - ansible_facts['lvm']['vgs']['first']['lvs'] | sort == ['one', 'two']
+          - ansible_facts['lvm']['vgs']['second']['lvs'] | sort == ['two', 'uno']
+          # only one lv named 'two' is represented in the top level lvs fact
+          - ansible_facts['lvm']['lvs']['two']['vg'] == 'second'
```
we should add test that extracts all lvs from vgs, uniques the names and compares to current/deprecated key

s-hertel committed:
Add warning if VG appears mid LVM fact gathering

Test unique LV names match between ansible_facts.lvm.vgs and ansible_facts.lvm.lvs

s-hertel committed:
Fix warning for unexpected scenario
